src/logika/GridGo.java

```diff
@@ -0,0 +1,280 @@
+			toReturn = false;
+		}
+		else if (prevPrevState == hash2D(grid)) {
+			toReturn = false; // grid repetition
+		}
+		this.placeColor(idx, FieldColor.EMPTY);
+		loadState(state);
+		return toReturn;
+	}
+	
+	private Map<Index, FieldColor> getZoneConrol() {
+		final int[][] NEIGHBOURS = new int[][] {{0, 1}, {1, 0}, {-1, 0}, {0, -1}};
+		
+		// Find all connected empty zones and all the stones bordering those zones.
+		Map<Index, Set<Index>> zonesNeighbours = new HashMap<Index, Set<Index>>();
+		Map<Index, Integer> zonesSizes = new HashMap<Index, Integer>();
+		for (Index idx : connectedComponents.getToplevels()) {
+			if (colorOfField(idx).equals(FieldColor.EMPTY)) {
+				zonesNeighbours.put(idx, new HashSet<Index>());
+				zonesSizes.put(idx, 0);
+			}
+		}
+		for (int i = 0; i < height; i++) {
+			for (int j = 0; j < width; j++) {
+				Index idx = new Index(i, j);
+				if (!colorOfField(idx).equals(FieldColor.EMPTY)) {
+					continue;
+				}
+				Index component = connectedComponents.getRepresentative(idx);
+				int oldSize = zonesSizes.get(component);
+				zonesSizes.put(component, oldSize + 1);
+				for (int didx = 0; didx < 4; didx++) {
+					int di = NEIGHBOURS[didx][0];
+					int dj = NEIGHBOURS[didx][1];
+					if (idx.i() + di < 0 || idx.j() + dj < 0 || idx.i() + di >= height || idx.j() + dj >= width)
+						continue;
+					Index neighbor = new Index(i + di, j + dj);
+					if (!colorOfField(neighbor).equals(FieldColor.EMPTY)) {
+						zonesNeighbours.get(connectedComponents.getRepresentative(idx)).add(neighbor);
+					}
+				}
+			}
+		}
+		
+		// Assign each empty zone a controller.
+		Map<Index, FieldColor> zoneControl = new HashMap<Index, FieldColor>();
+		for (Index zoneIdx : zonesNeighbours.keySet()) {
+			int black = 0, white = 0;
+			for (Index border : zonesNeighbours.get(zoneIdx)) {
+				if (colorOfField(border).equals(FieldColor.BLACK)) {
+					black++;
+				}
+				else {
+					assert colorOfField(border).equals(FieldColor.WHITE);
+					white++;
+				}
+			}
+			
+			double size = (double) zonesSizes.get(zoneIdx);
+			double blockage = (double) Integer.max(black, white);
+			double gridAverage = (double) (height + width) / 2.0;
+			
+			// The zone is neutral.
+			// Either both sides have the same claim to the zone or
+			// the second formula is correct. In the 9x9 case the formula means,
+			// that you need to have at least 1 stone per 7.2 controlled empty fields.
+			// The second case is here mostly to avoid black having control over the
+			// entire board every second turn at the start of the game.
+			if (black == white ||
+				size >= (blockage * gridAverage) * 0.8) {
+				zoneControl.put(zoneIdx, FieldColor.EMPTY);
+			}
+			else if (black > white) {
+				zoneControl.put(zoneIdx, FieldColor.BLACK);
+			}
+			else {
+				zoneControl.put(zoneIdx, FieldColor.WHITE);
+			}
+		}
+		
+		return zoneControl;
+	}
+	
+	private boolean isPrisoner(Index idx, Map<Index, FieldColor> zoneControl) {
```

Comment:
Kaj točno je algoritem tukaj? Kdaj polje šteješ kot ujetnika?

Reply:
Sem dodal komentar:
```java
/**
 * Is the stone at the given index a prisoner.
 * A stone is a prisoner if its component is a prisoner.
 * A component is a prisoner, if most of its liberties are
 * in zones controlled by the opponent.
 * @param idx The index of the field to be checked.
 * @param zoneControl Must be precalculated with getZoneControl 
 * for optimization purposes.
 * @return True iff the stone is a prisoner.
 */
```